src/Controls/src/Core/Platform/ModalNavigationManager/ModalNavigationManager.Android.cs

```diff
@@ -393,29 +391,56 @@ void FireAnimationEnded()
 				AnimationEnded?.Invoke(this, EventArgs.Empty);
 			}
 
-			public bool OnKey(IDialogInterface? dialog, [GeneratedEnum] Keycode keyCode, KeyEvent? e)
+			sealed class CustomComponentDialog : ComponentDialog
 			{
-				var mainActivity = Context?.GetActivity();
-				if (e is null || mainActivity is null)
+				public CustomComponentDialog(Context context, int themeResId) : base(context, themeResId)
 				{
-					return false;
+					this.OnBackPressedDispatcher.AddCallback(new CallBack(true, this));
 				}
 
-				return e.Action switch
+				public override bool OnKeyDown([GeneratedEnum] Keycode keyCode, KeyEvent? e)
 				{
-					KeyEventActions.Down => mainActivity.OnKeyDown(keyCode, e),
-					KeyEventActions.Up => mainActivity.OnKeyUp(keyCode, e),
-					KeyEventActions.Multiple => mainActivity.OnKeyMultiple(keyCode, e.RepeatCount, e),
-					_ => false,
-				};
-			}
+					var activity = Context?.GetActivity();
+					if (activity is null || e is null)
+						return false;
 
+					return activity.OnKeyDown(keyCode, e);
```

Comment:
Won't this cause side effects?
https://github.com/dotnet/maui/pull/30300#issuecomment-3016911896